vignettes/howto-dm-rows.Rmd

```diff
@@ -0,0 +1,241 @@
+
+Each method has its own requirements in order to maintain database consistency. 
+These involve constraints on primary key values as they are how rows are identified.
+
+| Method | Requirements |
+|--------|--------------|
+| dm_rows_insert() | The primary keys must differ from existing records.|
+| dm_rows_update() | Primary keys must match for all records to be updated.|
+| dm_rows_patch() | Updates missing values in existing records. Primary keys must match for all records to be patched.|
+| dm_rows_upsert() | Updates existing records and adds new records, based on the primary key.|
+| dm_rows_delete() | Removes matching records based on the primary key. The order in which the tables are processed is reversed.|
+| dm_rows_truncate() | Removes all records, only for tables in dm. Truncation is performed by truncating child tables before parent tables. |
+
+For `dm_rows_delete()` and `dm_rows_truncate()` tables linked to other tables via foreign keys are processed first. 
+For example, a database of nations with three linked tables, Country <- Province <- City.
+Foreign keys in `City` link rows to a row in `Province` and foreign keys in `Province` link rows to a row in `Country`.
+If it was discovered that Molvania, an imaginary country, was accidentally included in the database, on providing a dm object containing the `Country`, `Province` and `City` tables with all rows associated with it, the entries in the `City` table would be deleted first, followed by entries in the `Province` table, then its row in the `Country` table.
+This is to ensure the integrity of all relations during the deletion or truncation process.
+
+## Usage {#usage}
+
+To demonstrate the use of these table modifying methods we will create a simple dm object with two tables linked by a foreign key.
+
+Note the foreign key of `NA` in the `child` tibble.
+{dm} doesn't check your key values when you create a dm.
+You may want to use `dm_examine_constraints()` after it is created, but
+be aware that when using `dm_examine_constraints()` NULL (`NA`) foreign keys are allowed and will be counted as a match. 
+In some cases this doesn't make sense and non-NULL columns should be enforced by the RDBMS. 
+Currently {dm} does not specify or check non-NULL constraints for columns.
+
+``````{r }
+parent <- tibble(value=c('A', 'B', 'C'), pk=1:3)
+child <- tibble(value=c('a', 'b', 'c'), pk=1:3, fk=c(1,1,NA))
+
+demo_dm <- dm(parent=parent, child=child) %>%
+  dm_add_pk(parent,pk) %>%
+  dm_add_pk(child, pk) %>%
+  dm_add_fk(child, fk, parent)
+
+demo_dm
+dm_examine_constraints(demo_dm)
+``````
+
+Table modification methods work as expected on local dm objects, returning a new dm with the modifications in place.
+dm objects whose source is a database treat all operations as temporary unless you request explicitly to commit them by including the argument `temporary=FALSE`. 
+
+Below we copy `demo_dm` into an SQLite database.
```

Comment:
Can we start by copying the empty structure into the database and then adding the rows with `copy_dm_to(temporary = FALSE) %>% dm_rows_insert()` ? It's good practice to separate model creation from database population.